### DISCUSSION TITLE
Middleware-query reconciliation

### DISCUSSION THREAD

**Discussion by soryy708** [prisma/prisma#12909] [Q&A] [answered]
How do I reconcile between a middleware callback and a query event callback?

In the middleware I have the model and action, but not the SQL that they generate.
In the query event I have the SQL query, but not the model and action that generated it.

I'd like to trace and log both, but for that I need to create a relationship between the middleware and query event. How do I do that?

I observed that the event starts after the middleware starts, and ends before the middleware ends. I tried pushing the context to a queue/stack in the middleware and popping in the event, but this breaks if there is more than one concurrent action.

**Comment by austincrim** [comment]
Hey @soryy708 👋 ,

I don't think this is possible today, but I'm not totally sure. I'll pass this along to the engineering team.

Could you explain your use case a little more? That will help with any potential workarounds or feature requests.

**Reply by soryy708** [reply]
My use case: I'd like to send opentelemetry a span for the total time Prisma processed the query, including the model and action; and a child span for the total time the database processed the query, including the SQL statement.

**Reply by matthewmueller** [reply]
Hey @soryy708 that's actually exactly what we're working on with tracing: https://github.com/prisma/prisma/issues/9601. Emitting spans using OpenTelemetry.

We hope to have an early version out soon! We'll ping you when it's ready.

**Reply by soryy708** [reply]
Hi @matthewmueller . So I understand that currently there's no way to correlate a middleware callback its query event?

**Reply by soryy708** [reply]
Issue moved to https://github.com/prisma/client-planning/issues/21

**Reply by matthewmueller** [reply, marked as answer]
Hey @soryy708! Sorry for the late reply. Yes, with middleware it's not possible, but I wanted to share that we're finally shipping Prisma Client tracing!

Here's an early sneak peak of what's going out in our 4.2.0 release next Tuesday: https://deploy-preview-3408--prisma2-docs.netlify.app/docs/concepts/components/prisma-client/opentelemetry-tracing. You can actually use this today though since it's already available in the code since 4.1.0.

I'd love to get your feedback on our design and hit me up if you give tracing a shot in your app!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
